File: common/concept/Person.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.vaticle.typedb.benchmark.common.concept;

import javax.annotation.Nullable;
import java.time.LocalDateTime;
import java.util.Objects;

public class Person {

    private final String email;
    private final String firstName;
    private final String lastName;
    private final String address;
    private final Gender gender;
    private final LocalDateTime birthDate;
    private final int hash;

    public Person(String email) {
        this(email, null, null, null, null, null);
    }

    public Person(String email, @Nullable String firstName, @Nullable String lastName, @Nullable String address,
                  @Nullable Gender gender, @Nullable LocalDateTime birthDate) {
        this.email = email;
        this.firstName = firstName;
        this.lastName = lastName;
        this.address = address;
        this.gender = gender;
        this.birthDate = birthDate;
        hash = Objects.hash(email, firstName, lastName, address, gender, birthDate);
    }

    public String email() {
        return email;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Person that = (Person) o;
        return (email.equals(that.email) &&
                Objects.equals(firstName, that.firstName) &&
                Objects.equals(lastName, that.lastName) &&
                Objects.equals(address, that.address) &&
                Objects.equals(gender, that.gender) &&
                Objects.equals(birthDate, that.birthDate));
    }

    @Override
    public int hashCode() {
        return hash;
    }
}

File: common/concept/Person.kt
/*
 * Copyright (C) 2022 Vaticle
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.vaticle.typedb.benchmark.common.concept

import java.time.LocalDateTime
import java.util.Objects

class Person constructor(
    val email: String,
    private val firstName: String? = null,
    private val lastName: String? = null,
    private val address: String? = null,
    private val gender: Gender? = null,
    private val birthDate: LocalDateTime? = null
) {
    private val hash = Objects.hash(email, firstName, lastName, address, gender, birthDate)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as Person
        return email == that.email && firstName == that.firstName && lastName == that.lastName && address == that.address && gender == that.gender && birthDate == that.birthDate
    }

    override fun hashCode(): Int {
        return hash
    }
}
